go/client/chat_utils.go

```diff
@@ -2,84 +2,146 @@ package client
+
+	if timeStr := ctx.String("time"); len(timeStr) > 0 {
+		fetcher.selector.Since = &timeStr
 	}
 
-	var days int
-	if days, err = strconv.Atoi(s[:p]); err != nil {
-		return time.Duration(0), err
+	if ctx.Bool("all") {
+		fetcher.selector.Limit = 0
 	}
-	d = time.Duration(days) * 24 * time.Hour
 
-	if p < len(s) {
-		var dur time.Duration
-		if dur, err = time.ParseDuration(s[p+1:]); err != nil {
-			return time.Duration(0), err
-		}
-		d += dur
+	fetcher.resolver.TopicName = ctx.String("topic-name")
+	switch t := strings.ToLower(ctx.String("topic-type")); t {
+	case "chat":
+		fetcher.resolver.TopicType = chat1.TopicType_CHAT
+	case "dev":
+		fetcher.resolver.TopicType = chat1.TopicType_DEV
+	default:
+		err = fmt.Errorf("invalid topic-type %s. Has to be one of %v", t, []string{"chat", "dev"})
+		return fetcher, err
 	}
 
-	return d, nil
-}
+	fetcher.selector.MarkAsRead = markAsRead
+	fetcher.resolver.TlfName = tlfName
 
-func parseTimeFromRFC3339OrDurationFromPast(s string) (t time.Time, err error) {
-	var errt, errd error
-	var d time.Duration
+	return fetcher, nil
+}
 
-	if s == "" {
-		return
+func (f messageFetcher) fetch(ctx context.Context, g *libkb.GlobalContext) (messages cliChatMessages, err error) {
+	chatClient := f.chatClient // should be nil unless in test
+	if chatClient == nil {
+		chatClient, err = GetChatLocalClient(g)
+		if err != nil {
+			return nil, fmt.Errorf("Getting chat service client error: %s", err)
+		}
 	}
 
-	if t, errt = time.Parse(time.RFC3339, s); errt == nil {
-		return t, nil
+	conversationIDs, err := f.resolver.Resolve(ctx, chatClient)
+	if err != nil {
+		return nil, err
 	}
-	if d, errd = parseDurationExtended(s); errd == nil {
-		return time.Now().Add(-d), nil
+	// TODO: prompt user to choose conversation(s)
+	f.selector.Conversations = conversationIDs
+
+	conversations, err := chatClient.GetMessagesLocal(ctx, f.selector)
+	if err != nil {
+		return nil, fmt.Errorf("GetMessagesLocal error: %s", err)
 	}
 
-	return time.Time{}, fmt.Errorf("given string is neither a valid time (%s) nor a valid duration (%v)", errt, errd)
+	for _, conv := range conversations {
+		if len(conv.Messages) == 0 {
+			continue
+		}
+		m := conv.Messages[0]
+		var body string
+		switch t := m.MessagePlaintext.MessageBodies[0].Type; t {
+		case chat1.MessageType_TEXT:
+			body = formatChatText(m.MessagePlaintext.MessageBodies[0].Text)
+		case chat1.MessageType_ATTACHMENT:
+			body = formatChatAttachment(m.MessagePlaintext.MessageBodies[0].Attachment)
+		default:
+			g.Log.Debug("unsupported MessageType: %s", t)
+			continue
+		}
+
+		// TODO: Song: get rid of this in my next PR
+		cm := cliChatMessage{
+			with:          strings.Split(m.MessagePlaintext.ClientHeader.TlfName, ","),
+			timestamp:     gregor1.FromTime(m.ServerHeader.Ctime),
+			formattedBody: body,
+		}
+		if m.Info != nil {
+			cm.isNew = m.Info.IsNew
+			cm.author = m.Info.SenderUsername
+			cm.device = m.Info.SenderDeviceName
+			cm.topic = m.Info.TopicName
+		}
```

Comment:
can cliChatMessage just have an info field with m.Info instead of copying these?